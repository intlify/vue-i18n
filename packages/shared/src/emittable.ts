/**
 * Event type
 */
export type EventType = string | symbol

/**
 * Event handler
 */
export type EventHandler<T = unknown> = (payload?: T) => void

/**
 * Wildcard event handler
 */
export type WildcardEventHandler<T = Record<string, unknown>> = (
  event: keyof T,
  payload?: T[keyof T]
) => void

/**
 * Event handler list
 */
export type EventHandlerList<T = unknown> = Array<EventHandler<T>>

/**
 * Wildcard event handler list
 */
export type WildcardEventHandlerList<T = Record<string, unknown>> = Array<
  WildcardEventHandler<T>
>

/**
 * Event handler map
 */
export type EventHandlerMap<Events extends Record<EventType, unknown>> = Map<
  keyof Events | '*',
  EventHandlerList<Events[keyof Events]> | WildcardEventHandlerList<Events>
>

/**
 * Event emitter interface
 */
export interface Emittable<Events extends Record<EventType, unknown> = {}> {
  /**
   * A map of event names of registered event handlers
   */
  events: EventHandlerMap<Events>

  /**
   * Register an event handler with the event type
   *
   * @param event - An {@link EventType}
   * @param handler - An {@link EventHandler}, or a {@link WildcardEventHandler} if you are specified "*"
   */
  on<Key extends keyof Events>(
    event: Key | '*',
    handler: EventHandler<Events[keyof Events]> | WildcardEventHandler<Events>
  ): void

  /**
   * Unregister an event handler for the event type
   *
   * @param event - An {@link EventType}
   * @param handler - An {@link EventHandler}, or a {@link WildcardEventHandler} if you are specified "*"
   */
  off<Key extends keyof Events>(
    event: Key | '*',
    handler: EventHandler<Events[keyof Events]> | WildcardEventHandler<Events>
  ): void

  /**
   * Invoke all handlers with the event type
   *
   * @remarks
   * Note Manually firing "*" handlers should be not supported
   *
   * @param event - An {@link EventType}
   * @param payload - An event payload, optional
   */
  emit<Key extends keyof Events>(
    event: Key,
    payload?: Events[keyof Events]
  ): void
}
