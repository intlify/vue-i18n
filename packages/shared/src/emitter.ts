/**
 * Event emitter, forked from the below:
 * - original repository url: https://github.com/developit/mitt
 * - code url: https://github.com/developit/mitt/blob/master/src/index.ts
 * - author: Jason Miller (https://github.com/developit)
 * - license: MIT
 */

import type {
  EventType,
  EventHandler,
  WildcardEventHandler,
  EventHandlerList,
  WildcardEventHandlerList,
  EventHandlerMap,
  Emittable
} from './emittable'

/**
 * Create a event emitter
 *
 * @returns An event emitter
 */
export function createEmitter<
  Events extends Record<EventType, unknown>
>(): Emittable<Events> {
  type GenericEventHandler =
    | EventHandler<Events[keyof Events]>
    | WildcardEventHandler<Events>
  const events = new Map() as EventHandlerMap<Events>

  const emitter = {
    events,

    on<Key extends keyof Events>(
      event: Key | '*',
      handler: GenericEventHandler
    ): void {
      const handlers: Array<GenericEventHandler> | undefined = events.get(event)
      const added = handlers && handlers.push(handler)
      if (!added) {
        events.set(event, [handler] as EventHandlerList<Events[keyof Events]>)
      }
    },

    off<Key extends keyof Events>(
      event: Key | '*',
      handler: GenericEventHandler
    ): void {
      const handlers: Array<GenericEventHandler> | undefined = events.get(event)
      if (handlers) {
        handlers.splice(handlers.indexOf(handler) >>> 0, 1)
      }
    },

    emit<Key extends keyof Events>(
      event: Key,
      payload?: Events[keyof Events]
    ): void {
      ;((events.get(event) || []) as EventHandlerList<Events[keyof Events]>)
        .slice()
        .map(handler => handler(payload))
      ;((events.get('*') || []) as WildcardEventHandlerList<Events>)
        .slice()
        .map(handler => handler(event, payload))
    }
  }

  return emitter
}
