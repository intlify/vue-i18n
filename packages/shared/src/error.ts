export interface BaseError {
  code: number
}
