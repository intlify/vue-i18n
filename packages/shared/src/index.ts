/**
 * Original Utilities
 * written by kazuya kawaguchi
 */

export const inBrowser = typeof window !== 'undefined'

export let mark: (tag: string) => void | undefined
export let measure: (
  name: string,
  startTag: string,
  endTag: string
) => void | undefined

if (__DEV__) {
  const perf = inBrowser && window.performance
  if (
    perf &&
    perf.mark &&
    perf.measure &&
    perf.clearMarks &&
    perf.clearMeasures
  ) {
    mark = (tag: string): void => perf.mark(tag)
    measure = (name: string, startTag: string, endTag: string): void => {
      perf.measure(name, startTag, endTag)
      perf.clearMarks(startTag)
      perf.clearMarks(endTag)
    }
  }
}

const RE_ARGS = /\{([0-9a-zA-Z]+)\}/g

/* eslint-disable */
export function format(message: string, ...args: any): string {
  if (args.length === 1 && isObject(args[0])) {
    args = args[0]
  }
  if (!args || !args.hasOwnProperty) {
    args = {}
  }
  return message.replace(
    RE_ARGS,
    (match: string, identifier: string): string => {
      return args.hasOwnProperty(identifier) ? args[identifier] : ''
    }
  )
}

const hasSymbol =
  typeof Symbol === 'function' && typeof Symbol.toStringTag === 'symbol'

export const makeSymbol = (name: string): symbol | string => hasSymbol ? Symbol(name) : name

export const generateFormatCacheKey = (
  locale: string,
  key: string,
  source: string
): string => friendlyJSONstringify({ l: locale, k: key, s: source })

export const friendlyJSONstringify = (json: unknown): string =>
  JSON.stringify(json)
    .replace(/\u2028/g, '\\u2028')
    .replace(/\u2029/g, '\\u2029')
    .replace(/\u0027/g, '\\u0027')

export const isNumber = (val: unknown): val is number =>
  typeof val === 'number' && isFinite(val)

export const isDate = (val: unknown): val is Date =>
  toTypeString(val) === '[object Date]'

export const isRegExp = (val: unknown): val is RegExp =>
  toTypeString(val) === '[object RegExp]'

export const isEmptyObject = (val: unknown): val is boolean =>
  isPlainObject(val) && Object.keys(val).length === 0

export function warn(msg: string, err?: Error): void {
  if (typeof console !== 'undefined') {
    console.warn(`[intlify] ` + msg)
    /* istanbul ignore if */
    if (err) {
      console.warn(err.stack)
    }
  }
}

let _globalThis: any
export const getGlobalThis = (): any => {
  // prettier-ignore
  return (
    _globalThis ||
    (_globalThis =
      typeof globalThis !== 'undefined'
        ? globalThis
        : typeof self !== 'undefined'
          ? self
          : typeof window !== 'undefined'
            ? window
            : typeof global !== 'undefined'
              ? global
              : {})
  )
}

export function escapeHtml(rawText: string): string {
  return rawText
    .replace(/</g, '&lt;')
    .replace(/>/g, '&gt;')
    .replace(/"/g, '&quot;')
    .replace(/'/g, '&apos;')
}

const hasOwnProperty = Object.prototype.hasOwnProperty
export function hasOwn(obj: object | Array<any>, key: string): boolean {
  return hasOwnProperty.call(obj, key)
}

/* eslint-enable */

/**
 * Useful Utilities By Evan you
 * Modified by kazuya kawaguchi
 * MIT License
 * https://github.com/vuejs/vue-next/blob/master/packages/shared/src/index.ts
 * https://github.com/vuejs/vue-next/blob/master/packages/shared/src/codeframe.ts
 */
export const isArray = Array.isArray
export const isFunction = (val: unknown): val is Function =>
  typeof val === 'function'
export const isString = (val: unknown): val is string => typeof val === 'string'
export const isBoolean = (val: unknown): val is boolean =>
  typeof val === 'boolean'
export const isSymbol = (val: unknown): val is symbol => typeof val === 'symbol'
export const isObject = (val: unknown): val is Record<any, any> => // eslint-disable-line
  val !== null && typeof val === 'object'

export const isPromise = <T = any>(val: unknown): val is Promise<T> => { // eslint-disable-line
  return isObject(val) && isFunction(val.then) && isFunction(val.catch)
}

export const objectToString = Object.prototype.toString
export const toTypeString = (value: unknown): string =>
  objectToString.call(value)

export const isPlainObject = (val: unknown): val is object =>
  toTypeString(val) === '[object Object]'

// for converting list and named values to displayed strings.
export const toDisplayString = (val: unknown): string => {
  return val == null
    ? ''
    : isArray(val) || (isPlainObject(val) && val.toString === objectToString)
    ? JSON.stringify(val, null, 2)
    : String(val)
}

const RANGE = 2

export function generateCodeFrame(
  source: string,
  start = 0,
  end = source.length
): string {
  const lines = source.split(/\r?\n/)
  let count = 0
  const res: string[] = []
  for (let i = 0; i < lines.length; i++) {
    count += lines[i].length + 1
    if (count >= start) {
      for (let j = i - RANGE; j <= i + RANGE || end > count; j++) {
        if (j < 0 || j >= lines.length) continue
        const line = j + 1
        res.push(`${line}${' '.repeat(3 - String(line).length)}|  ${lines[j]}`)
        const lineLength = lines[j].length
        if (j === i) {
          // push underline
          const pad = start - (count - lineLength) + 1
          const length = Math.max(
            1,
            end > count ? lineLength - pad : end - start
          )
          res.push(`   |  ` + ' '.repeat(pad) + '^'.repeat(length))
        } else if (j > i) {
          if (end > count) {
            const length = Math.max(Math.min(end - count, lineLength), 1)
            res.push(`   |  ` + '^'.repeat(length))
          }
          count += lineLength + 1
        }
      }
      break
    }
  }
  return res.join('\n')
}
