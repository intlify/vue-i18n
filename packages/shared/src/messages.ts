import { create, isArray, isObject } from './utils'

const isNotObjectOrIsArray = (val: unknown) => !isObject(val) || isArray(val)
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function deepCopy(src: any, des: any): void {
  // src and des should both be objects, and none of them can be a array
  if (isNotObjectOrIsArray(src) || isNotObjectOrIsArray(des)) {
    throw new Error('Invalid value')
  }

  const stack = [{ src, des }]
  while (stack.length) {
    const { src, des } = stack.pop()!

    // using `Object.keys` which skips prototype properties
    Object.keys(src).forEach(key => {
      if (key === '__proto__') {
        return
      }
      // if src[key] is an object/array, set des[key]
      // to empty object/array to prevent setting by reference
      if (isObject(src[key]) && !isObject(des[key])) {
        des[key] = Array.isArray(src[key]) ? [] : create()
      }

      if (isNotObjectOrIsArray(des[key]) || isNotObjectOrIsArray(src[key])) {
        // replace with src[key] when:
        // src[key] or des[key] is not an object, or
        // src[key] or des[key] is an array
        des[key] = src[key]
      } else {
        // src[key] and des[key] are both objects, merge them
        stack.push({ src: src[key], des: des[key] })
      }
    })
  }
}
