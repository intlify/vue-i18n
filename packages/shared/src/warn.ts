export function warn(msg: string, err?: Error): void {
  if (typeof console !== 'undefined') {
    console.warn(`[intlify] ` + msg)
    /* istanbul ignore if */
    if (err) {
      console.warn(err.stack)
    }
  }
}

const hasWarned: Record<string, boolean> = {}

export function warnOnce(msg: string) {
  if (!hasWarned[msg]) {
    hasWarned[msg] = true
    warn(msg)
  }
}
