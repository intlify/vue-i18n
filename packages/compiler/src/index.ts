import { isObject } from '@intlify/shared'

export type Foo = 'Foo'

console.log('hello compiler!', isObject({}))
