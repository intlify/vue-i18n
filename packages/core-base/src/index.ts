import { getGlobalThis } from '@intlify/shared'

export {
  CompileError,
  CompileErrorCodes,
  createCompileError
} from '@intlify/message-compiler'
export * from './resolver'
export * from './runtime'
export * from './context'
export * from './fallbacker'
export * from './compile'
export * from './translate'
export * from './datetime'
export * from './number'
export { getWarnMessage, CoreWarnCodes } from './warnings'
export { CoreError, CoreErrorCodes, createCoreError } from './errors'
export * from './types'
export * from './devtools'

if (__ESM_BUNDLER__ && !__TEST__) {
  if (typeof __FEATURE_PROD_INTLIFY_DEVTOOLS__ !== 'boolean') {
    getGlobalThis().__INTLIFY_PROD_DEVTOOLS__ = false
  }
}
