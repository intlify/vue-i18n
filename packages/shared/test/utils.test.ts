import { format, generateCodeFrame, makeSymbol, join } from '../src/index'

test('format', () => {
  expect(format(`foo: {0}`, 'x')).toEqual('foo: x')
  expect(format(`foo: {0}, {1}`, 'x', 'y')).toEqual('foo: x, y')
  expect(format(`foo: {x}, {y}`, { x: 1, y: 2 })).toEqual('foo: 1, 2')
})

test('generateCodeFrame', () => {
  const source = `hi, { 'kazupon' }`.trim()
  const keyStart = source.indexOf(`{ 'kazupon' }`)
  const keyEnd = keyStart + `{ 'kazupon' }`.length
  expect(generateCodeFrame(source, keyStart, keyEnd)).toMatchSnapshot()
})

test('makeSymbol', () => {
  expect(makeSymbol('foo')).not.toEqual(makeSymbol('foo'))
  expect(makeSymbol('bar', true)).toEqual(makeSymbol('bar', true))
})

test('join', () => {
  expect(join([])).toEqual([].join(''))
  expect(join(['a'], ',')).toEqual(['a'].join(','))
  expect(join(['a', 'b', 'c'])).toEqual(['a', 'b', 'c'].join(''))
  expect(join(['a', 'b', 'c'], ' ')).toEqual(['a', 'b', 'c'].join(' '))

  const longSize = [
    'a',
    'b',
    'c',
    'd',
    'e',
    'f',
    'g',
    'h',
    'i',
    'j',
    'k',
    'l',
    'm',
    'n',
    'o',
    'p',
    'q',
    'r',
    's',
    't',
    'u',
    'v',
    'w',
    'x',
    'y',
    'z'
  ]
  expect(join(longSize, ' ')).toEqual(longSize.join(' '))
})
