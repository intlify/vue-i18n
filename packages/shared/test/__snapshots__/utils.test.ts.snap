// Vitest Snapshot v1

exports[`generateCodeFrame 1`] = `
"1  |  hi, { 'kazupon' }
   |      ^^^^^^^^^^^^^"
`;
